### PR TITLE
doc: added importance and custom markers section

### PR DESCRIPTION
**Pull request by sidecontrol**

### CODE REVIEW

**Comment by jakub-vavra-cz** [comment]
Hello, looks mostly good.
We might want to add also "security" marker from the start and label tests that cover logins with wrong password, expired/disabled accounts eg. . WDYIT @pbrezina?

**Review by jakub-vavra-cz — APPROVED** [review]
LGTM

**Comment by sidecontrol** [comment]
@jakub-vavra-cz , some of it should be under authentication, and we don't have any access control tests ported in yet, so that group will be added eventually.

File: docs/marks.rst [hunk 1]
```diff
@@ -113,6 +113,33 @@ You can then run tests that are relevant only to the selected ticket:
     cd src/tests/system
     pytest --mh-config=mhc.yaml --mh-lazy-ssh -v --ticket=gh#3433
 
+@pytest.mark.importance
+=======================
```
Comment (pbrezina) [inline docs/marks.rst:117]: I'm missing a guide how to clasify the tests. What makes the test "critical", what makes it "low"?

File: docs/marks.rst [hunk 2]
```diff
@@ -132,7 +159,7 @@ The ``@pytest.mark.tier`` takes single number as an argument.
     def test_tier1()
         pass
 
-You can then run tests that are relevant only to the selected ticket:
+You can then run tests by tier:
```
Comment (pbrezina) [inline docs/marks.rst:162]: Tier section can be completely removed as we switched from tier to importance.

File: docs/marks.rst (hunk 1 above)
Comment (pbrezina) [reply docs/marks.rst:117]: How is QA using this marker? It doesn't really make sense form upstream point of view, so this should be also clarified.

**Review by pbrezina — CHANGES_REQUESTED** [review]

**Comment by pbrezina** [comment]
> Hello, looks mostly good. We might want to add also "security" marker from the start and label tests that cover logins with wrong password, expired/disabled accounts eg. . WDYIT @pbrezina?

AFAIK this is the "authentication" or "authorization" marker. So my opinion is to either use these, or scratch them in favor of "security".

---

From my upstream point of view, both importance and custom markers do not really sense. Before release and before pull request push, all tests must be green. So IIUIC it is more related to how you want to run the tests in downstream and this should be explained here so however reads it knows the meaning of these markers

File: docs/marks.rst (hunk 2 above)
Comment (sidecontrol) [reply docs/marks.rst:162]: AFAIK, tiers will be designated for different architectures. When we run our regression runs, tier0 will be x86_64, tier1, tier2 will be other architectures. For now, I'll remove it, but it may be added again later.

**Comment by sidecontrol** [comment]
@pbrezina, it's simply a way to organize the tests when we execute regression runs, so we can run a subset of tests that a patch, build, and errata effects without executing all tests. Gating, for example, gating is supposed to be a  quick, core tests that touch all features, aka "critical". While "high" will be a comprehensive run of the features. Medium will be features that are not used all the time or will not impact customer's operations if they fail, like the CLI tools,  sssctl, sss_override. 

I'll expand the documentation and add this kind of thing to the doc.

**Comment by pbrezina** [comment]
> @pbrezina, it's simply a way to organize the tests when we execute regression runs, so we can run a subset of tests that a patch, build, and errata effects without executing all tests. Gating, for example, gating is supposed to be a quick, core tests that touch all features, aka "critical". While "high" will be a comprehensive run of the features. Medium will be features that are not used all the time or will not impact customer's operations if they fail, like the CLI tools, sssctl, sss_override.

Yes, exactly this kind of thing should go to the docs.

> 
> I'll expand the documentation and add this kind of thing to the doc.

**Comment by jakub-vavra-cz** [comment]
@pbrezina, @sidecontrol  as for difference between critical, high, medium, low.
The categorization of tests is important internally for dividing test to tiers (that will have a different frequence of execution), it is useful to evaluate seriousness of possible failure. (Failure of low importance test will have a different impact that critical test.  

"Critical" tests should contain only small sample of tests for the most important functionality.
These should have very short runtime so they can be run often (like on PRs).
They should be covering only the "sunny day" (positive test) scenarios.

"High" tests that cover all functionality properly still focusing on positive scenarios.
Consider tests with properly configured sssd and kerberos, automount, ... making sure that everything works properly.

"Medium" tests comprehensively including some less used functionality and more complex setups.
Tests with slightly misconfigured sssd, less common setups, basic failover.

"Low" tests for a rarely used functionality, edge cases and very specific complex setups, performance/stress tests.
Tests where the sssd / environment is misconfigured.

**Comment by sidecontrol** [comment]
@jakub-vavra-cz , how does this sound?

    * - critical
      - Core subset of tests that covers all operational features. This is used for gating where it maybe ran several
        times a day. To manage resources, the execution all these tests should be kept under an hour.
    * - high
      - The most comprehensive set of tests, that covers all operational features.
    * - medium
      - Tess that do not impact operational functionality, like the CLI commands included in sss-tools package.
    * - low
      - Tests that may have a long execution time, edge cases or complex scenarios that demand a lot of resources.

**Comment by jakub-vavra-cz** [comment]
Hi @sidecontrol , 
I would shift some stuff from critical to high. Gating is Crit+High, I would keep Critical for smoke tests that we will execute 
on commits/PRs.

* - critical
  - Core subset of tests that covers most important operational features. This is used in pipelines where it maybe ran multiple times a day. The execution should be kept as short as possible.
 * - high
   - The comprehensive set of tests, that covers all operational features.  This is used for gating where it maybe ran several  times a day. To manage resources, the execution all these tests should be kept under an hour.
 * - medium
    - Extended set that covers tests that do not impact operational functionality, like the CLI commands included in sss-tools package. Tests that cover negative scenarios and misconfigured environment fit here as well.
 * - low
   - Tests that may have a long execution time, edge cases or complex scenarios that demand a lot of resources. Consider performance and stress tests as part of this set.

**Comment by sidecontrol** [comment]
Updated.

**Review by jakub-vavra-cz — DISMISSED** [review]
LGTM

File: docs/marks.rst [hunk 3]
```diff
@@ -113,42 +113,101 @@ You can then run tests that are relevant only to the selected ticket:
...
-@pytest.mark.tier
-=================
+@pytest.mark.importance
+=======================
 
-The `tier mark <https://github.com/next-actions/pytest-tier>`__ can
-associate a test with a specific tier.
+The `importance mark <https://github.com/next-actions/pytest-importance>`__ can
+associate a test with a level of importance. This is used by quality engineers to
+prioritize and group the test determined by the level of impact to the customer.
```
Comment (pbrezina) [inline docs/marks.rst:121]: "and group the test**s**"?

File: docs/marks.rst [hunk 4]
```diff
@@ -113,42 +113,101 @@ You can then run tests that are relevant only to the selected ticket:
...
 
-The ``@pytest.mark.tier`` takes single number as an argument.
+The ``@pytest.mark.importance`` takes a string as an argument. The values used
+are "critical", "high", "medium" and "low".
```
Comment (pbrezina) [inline docs/marks.rst:124]: If no importance marker is given, it defaults to "medium".

File: docs/marks.rst [hunk 5]
```diff
@@ -113,42 +113,101 @@ You can then run tests that are relevant only to the selected ticket:
...
     :stub-columns: 1
 
-    * - Tier
+    * - Importance
       - Description
-    * - @TODO
-      -
+    * - critical
+      - Core subset of tests that covers most important operational features. This is used in pipelines where it maybe
+        ran multiple times a day. The execution should be kept as short as possible.
```
Comment (pbrezina) [inline docs/marks.rst:156]: where it maybe ran -> where it can be ran?

File: docs/marks.rst [hunk 6]
```diff
@@ -113,42 +113,101 @@ You can then run tests that are relevant only to the selected ticket:
...
+    * - high
+      - The comprehensive set of tests, that covers all operational features. This is used for gating where it maybe
+        ran several times a day. To manage resources, the execution all these tests should be kept under an hour.
```
Comment (pbrezina) [inline docs/marks.rst:159]: where it maybe ran -> where it can be ran?

File: docs/marks.rst [hunk 7]
```diff
@@ -113,42 +113,101 @@ You can then run tests that are relevant only to the selected ticket:
...
+    pytest --mh-config=mhc.yaml --mh-lazy-ssh -v --importance="critical"
 
-Tier definition
-===============
+Importance definition
+=====================
 
-.. list-table:: Tier definition
-    :align: center
+.. list-table::
```
Comment (pbrezina) [inline docs/marks.rst:147]: I would avoid using table and just do it in bullet points. The description is long and it does not show up nicely in the table since it does not support multiline fields.

```rst
* **critical**: description ...
```

Same for the custom markers.

**Review by pbrezina — CHANGES_REQUESTED** [review]
Thanks, see minor comments inside.

File: docs/marks.rst (hunk 3 above)
Comment (sidecontrol) [reply docs/marks.rst]: "A group of critical tests", but it's implied, removing it.

File: docs/marks.rst [hunk 8]
```diff
@@ -113,42 +113,83 @@ You can then run tests that are relevant only to the selected ticket:
...
+You can then run the tests by importance:
 
 .. code-block:: text
 
     cd src/tests/system
-    pytest --mh-config=mhc.yaml --mh-lazy-ssh -v --tier=1
+    pytest --mh-config=mhc.yaml --mh-lazy-ssh -v --importance="critical"
+
+Importance definition
+=====================
```
Comment (pbrezina) [inline docs/marks.rst:146]: This is a subsection, you `----` instead of `====`.

File: docs/marks.rst [hunk 9]
```diff
@@ -113,42 +113,83 @@ You can then run tests that are relevant only to the selected ticket:
...
-      -
+You can run tests by custom markers:
+
+.. code-block:: text
+
+    cd src/tests/system
+    pytest --mh-config-mhc=mhc.yaml --mh-lazy-ssh -v -m authentication
+
+Custom Marker definitions
+=========================
```
Comment (pbrezina) [inline docs/marks.rst:186]: This is a subsection, you `----` instead of `====`.

**Review by pbrezina — CHANGES_REQUESTED** [review]

**Comment by patriki01** [comment]
Hi @sidecontrol,
It seems that you forgot to put `config` marker to your original [PR to Testing Framework](https://github.com/SSSD/sssd/pull/6902). Jakub has already added the [marker](https://github.com/SSSD/sssd/pull/6955). Could you please add it to the docs as well?
Thanks

**Comment by sidecontrol** [comment]
@patriki01  Done.

**Review by jakub-vavra-cz — APPROVED** [review]
LGTM

**Review by pbrezina — APPROVED** [review]
